api/envoy/extensions/http/original_ip_detection/xff/v3/xff.proto

```diff
@@ -22,5 +24,49 @@ message XffConfig {
   // determining the origin client's IP address. The default is zero if this option
   // is not specified. See the documentation for
   // :ref:`config_http_conn_man_headers_x-forwarded-for` for more information.
+  //
+  // Only one of ``xff_num_trusted_hops`` and ``xff_trusted_cidrs`` can be set.
   uint32 xff_num_trusted_hops = 1;
+
+  // The `CIDR <https://tools.ietf.org/html/rfc4632>`_ ranges to trust when
+  // evaluating the remote IP address to determine the origin client's IP address.
+  // This is used instead of
+  // :ref:`use_remote_address <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpConnectionManager.use_remote_address>`.
+  // When the remote IP address matches a trusted CIDR and the
+  // :ref:`config_http_conn_man_headers_x-forwarded-for` header was sent, the last
+  // IP address from ``x-forwarded-for`` is used as the original client address.
+  // If no ``x-forwarded-for`` header was sent, the remote IP is considered the real IP.
+  //
+  // If ``recurse`` is set the :ref:`config_http_conn_man_headers_x-forwarded-for`
+  // header is also evaluated against the trusted CIDR list, and the last non-trusted
+  // address is used as the original client address. If all addresses in ``x-forwarded-for``
+  // are within the trusted list, the first entry is used.
+  //
+  // This is typically used when requests are proxied by a
+  // `CDN <https://en.wikipedia.org/wiki/Content_delivery_network>`_.
+  //
+  // Only one of ``xff_num_trusted_hops`` and ``xff_trusted_cidrs`` can be set.
+  //
+  // Note: Use of this option is an ``n*m`` operation on every request, where ``n`` is
+  // the number of trusted CIDRs and ``m`` is the number of entries in the XFF header.
+  XffTrustedCidrs xff_trusted_cidrs = 2;
+
+  // Whether to append the detected trusted client IP address to ``x-forwarded-for``.
+  bool append_xff = 3;
```

Comment:
Sorry for being dense and asking this now, but don't we already append to XFF somewhere in the request pipeline? Does this mean that XFF will get double appended in some cases? If this is somehow different can you spell out how this is different and when this would be used?

Reply:
HCM will append to XFF only when `use_remote_address` is true: https://github.com/envoyproxy/envoy/blob/bc0190146f08458ac5a58af5b3545b24660fa8f8/source/common/http/conn_manager_utility.cc#L136-L143

When an original IP detection extension is used instead - which is incompatible with `use_remote_address` we never hit that code path, so there was no way to append to XFF from an original IP extension before this change.

Reply:
I see OK makes sense, thanks.

Reply:
maybe worth commenting in the API docs, as I had the same question.   "this only applies when use_remote_address is false, otherwise [existing config field] applies"